### ISSUE TITLE
Julep: lowering of keyword-function methods

### ISSUE THREAD

**Issue by timholy** [JuliaLang/julia#30908]
## TL;DR version

Proposal: for the "body method" produced when lowering kwarg functions, don't gensym the method name. For example, when `foo` has one or more keyword arguments, use `#foo#` instead of `#foo#37`. 

If you're not yet convinced, then the following tome might help.

## Background

The current situation makes it hard to compute valid signatures for all methods defined in a file. This turns out to be problematic for several purposes, most notably for debuggers (both Rebugger and ASTInterpreter2):

1. IDE debuggers need to be able to navigate to the first line of a method so the user can watch as it steps through the source code line-by-line.
1. when using a debugger, there's a temptation to edit the source code to, e.g., fix bugs
1. editing the code can confuse the debugger, because it changes the line numbering of the file. The risk is that when you step into `foo`, the editor (following the debugger's request to present file "filename.jl" at line 555) might land you halfway through the middle of an unrelated method because you deleted a few comments higher up in the file.
1. a solution is available in Revise, which tries to fix this by maintaining a dict of "corrected location" info indexed by the method's signature. The signature allows you to uniquely identify a method, so you can go from methods (obtained from stacktraces or by `which` when you want to `step in`) to the correct current position in the source file.

Hence, it's important to be able to parse a source file and generate an `IdDict{Signature,LocationInfo}` containing all signatures.

## The problem

Parsing signatures is generally easy using the lowered representation of a source file. An annoying exception occurs for functions with keyword arguments. Consider the following:

```julia
julia> module Lowering end
Main.Lowering

julia> ex = :(foo(x::Int8, y=0; z::String="hello") = x+y)
:(foo(x::Int8, y=0; z::String="hello") = begin
          #= REPL[2]:1 =#
          x + y
      end)

julia> Core.eval(Lowering, ex)
foo (generic function with 2 methods)

julia> names(Lowering; all=true)
10-element Array{Symbol,1}:
 Symbol("##foo#1") 
 Symbol("#eval")   
 Symbol("#foo")    
 Symbol("#foo#1")  
 Symbol("#include")
 Symbol("#kw##foo")
 :Lowering         
 :eval             
 :foo              
 :include          

julia> Meta.lower(Lowering, ex)
:($(Expr(:thunk, CodeInfo(
1 ─       $(Expr(:method, Symbol("#foo#2")))
│   %2  = (Core.typeof)(#foo#2)
│   %3  = (Core.svec)(%2, String, Core.Any, Int8, Core.Any)
│   %4  = (Core.svec)()
│   %5  = (Core.svec)(%3, %4)
│         $(Expr(:method, Symbol("#foo#2"), :(%5), CodeInfo(quote
    x + y
    return %1
end)))
...  # many more methods
```

The issue here is that pesky `#foo#2`, the name given to what here I'll refer to as the "body method" (it's what implements the actual body of `foo`). 

The problem because apparent when you note that what's present as a name in `Lowering` is `#foo#1`. Because the function's type is in the signature, we have a problem:

```julia
julia> fhash = getfield(Lowering, Symbol("#foo#1"))
#foo#1 (generic function with 1 method)

julia> meth = first(methods(fhash))
#foo#1(z::String, ::Any, x::Int8, y) in Main.Lowering at REPL[2]:1

julia> meth.sig
Tuple{getfield(Main.Lowering, Symbol("##foo#1")),String,Any,Int8,Any}
```

But if we naively compute the signature of the "body method" by just stepping through the lowered code, we'll compute

```julia
 Tuple{getfield(Main.Lowering, Symbol("##foo#2")),String,Any,Int8,Any}
```

This has two problems: `typeof(#foo#2)` doesn't exist (we didn't `eval` the lowered code), so you can't even compute this type. But even if you could, the debugger wouldn't be able to find the corrected source line by indexing `locationdict[sig]`: imagine `foo` throws an error and we want to stop the debugger there, the `sig` of the method in the stack trace is for `#foo#1` but `locationdict` only has a signature for `#foo#2`. 

There *is* a way to compute the correct signature for this method:

1. grab the symbol, note it's a gensym and that you need to delay processing of the signature.
1. scan forward in the lowered code until you get to a 3-arg `:method` Expr who's `arg[3].code[1]` is a `:call` expression with `arg[1]` that isn't a `SlotNumber` (`SlotNumber` indicates `#self#` calls and is used for filling defaults of positional arguments). For this example, this would be `foo(::Int8, ::Any)`---this method handles calls with complete positional arguments and no supplied keywords. 
1. In the lowered code, the body of this method is `#foo#2("hello", #self#, x, y); return %1`. That's no good. But we can use the lowered code to extract its signature `foo(::Int8, ::Any)` and then call `which`. This will give us the `eval`ed variant of this method. Assign this to a variable called `methnokw`.
1. Call `Base.uncompressed_ast(methnokw)` to retrieve the lowered code of its body, which is `#foo#1("hello", #self#, x, y); return %1`. From this we can extract the `#foo#1` we needed all along.
1. Jump back up to the top of the lowered code for `foo`, replace `#foo#2` with `#foo#1`, and generate the signature.

All of that is achievable, but it's nasty enough that I wonder if we should discuss making it more straightforward.

(One final detail is that for `UnionAll` signatures you need to look up the "active" signature by reflection. The issue is that two `UnionAll`s constructed with different `TypeVar`s---one from the lowered representation and one when the method definition was last `eval`ed---don't compare as equal. But that's not specific to kwarg functions and is much easier to handle.)

## A possible solution

The solution seems easy: why does this method need to be `gensym`med? Can we just use a consistent name `#foo#`? If we re-evaluate `foo` we'd probably prefer to overwrite the body method anyway, right? Currently if you re-eval `foo` many times you end up with

 ```julia
julia> names(Lowering; all=true)
20-element Array{Symbol,1}:
 Symbol("##foo#1") 
 Symbol("##foo#4") 
 Symbol("##foo#5") 
 Symbol("##foo#6") 
 Symbol("##foo#7") 
 Symbol("##foo#8") 
 Symbol("#eval")   
 Symbol("#foo")    
 Symbol("#foo#1")  
 Symbol("#foo#4")  
 Symbol("#foo#5")  
 Symbol("#foo#6")  
 Symbol("#foo#7")  
 Symbol("#foo#8")  
 Symbol("#include")
 Symbol("#kw##foo")
 :Lowering         
 :eval             
 :foo              
 :include          
```

and I don't see a strong reason for having all those `#foo#n` methods.

**Comment by jekbradbury** [comment]
This might make it easier for Cassette and friends to overdub kwargs methods too.

**Comment by simonbyrne** [comment]
See https://github.com/jrevels/Cassette.jl/issues/48 for the Cassette issue, though I'm not sure if what is proposed here would be a fix.

I'm not sure if it would be possible, but an alternative is to have a wrapper type, say
```julia
struct Kwcall{F}
    f::F
end
```
so that
```julia
(f::F)(args...;kwargs...)
```
lowers to
```julia
Kwcall{F}(f)(kwargs, args...)
```

In that way Cassette could just define:
```julia
Cassette.overdub(ctx::Cassette.Context, kw::Kwcall, kwargs, args...) = 
    Cassette.overdub(ctx, kw.f, args...; kwargs...)
```

**Comment by timholy** [comment]
The Cassette issue references a slightly different method (this terrain is complex...): the `#kw##foo` method handles assignment of default values when some but perhaps not all kwargs are specified, and also putting the kw arguments into a consistent order. It's accessible like this:

```julia
julia> methods(Core.kwfunc(Lowering.foo))
# 2 methods for generic function "#foo":
[1] (::getfield(Main.Lowering, Symbol("#kw##foo")))(::Any, ::typeof(Main.Lowering.foo), x::Int8) in Main.Lowering
[2] (::getfield(Main.Lowering, Symbol("#kw##foo")))(::Any, ::typeof(Main.Lowering.foo), x::Int8, y) in Main.Lowering
```

These get called (via lowering) anytime you call `foo` and specify at least one kwarg. The last two lines of this dispatch chain are, e.g.,

```julia
    #foo#2(z, , x, y)
    return %21
```

The "empty argument" (printed this way because the corresponding `codeinfo.slotname` is `""`) gets filled (IIUC) with `#self#`, meaning that `foo` passes itself as an argument so that its "heritage" is clear. Ultimately the kw function results in a call `#foo#n(kwvalues_in_order..., foo, positional_args...)`. It's the `#foo#n` method that does the actual work of the method's body, and it's the only method with a (non-reproducible) `gensym`med name. In contrast, the methods that fill in default kw values *always* have the name `#kw##foo`).

You can see the dispatch chain by changing the definition of `foo` as
```julia
ex = :(foo(x::Int8, y=0; z::String="hello") = error("whoops"))
```

so that from the stack traces you see
```julia
julia> Lowering.foo(Int8(1))
ERROR: whoops
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] #foo#4(::String, ::Function, ::Int8, ::Int64) at ./REPL[12]:1
 [3] foo(::Int8, ::Int64) at ./REPL[12]:1 (repeats 2 times)
 [4] top-level scope at none:0

julia> Lowering.foo(Int8(1); z="hey")
ERROR: whoops
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] #foo#4(::String, ::Function, ::Int8, ::Int64) at ./REPL[12]:1
 [3] (::getfield(Main.Lowering, Symbol("#kw##foo")))(::NamedTuple{(:z,),Tuple{String}}, ::typeof(Main.Lowering.foo), ::Int8, ::Int64) at ./none:0 (repeats 2 times)
 [4] top-level scope at none:0
```

In the first case it never needs to call the `#kw##foo` method, because we didn't specify any kwargs so it's easy to fill in all the defaults in order and go straight to the "real method" which is now being called `#foo#4`. In the second case, we specify one kwarg, in which case the `#kw##foo` method is needed to fill in the rest (which don't exist in this case, but that's just a detail...).

**Comment by JeffBezanson** [comment]
I think this could probably work for top-level functions, but the question is what to do about closures and methods defined by type, e.g. `(f::Something)(x; kw...) = ...`. A consistent naming scheme is not sufficient for that. I would be open to adding some kind of metadata, e.g. a field of `Method`, that tells you the "body" method for a given method.

**Comment by timholy** [comment]
I thought of that, but I'm not certain it's a panacea. When you get the signature for `#foo#2`, which args come from the keywords and which are positional? Without knowing that, how do you construct

```julia
nonbodymeth = Tuple{getfield(mod, :foo), Int8, Any}
bodymeth = Core.kwbody(nonbodymeth)
```
?

**Comment by timholy** [comment]
This issue now has a workaround: https://github.com/JuliaDebug/LoweredCodeUtils.jl.
Initially this was internal to my local branch of the in-progress Revise rewrite, but since this operation is pretty complicated---and potentially useful to others---I thought it might be better to make it standalone.

**Comment by ronisbr** [comment]
Hi! Just to add an information, I would love to have some way to return the body function in those cases. I am trying to improve the initialization time of a package, and providing precompile statements for those body functions helps a lot. For now, I am using the function [here](https://github.com/juliamatlab/MatLang/blob/070a4fc31f4ddc4c691ad69c551776bca75bf37f/deps/SnoopCompile/precompile/apple/1.4/precompile_MatLang.jl#L9) to obtain the name of the body function, but I think it will be very good to have a native way to do so.

**Comment by timholy** [comment]
Yeah, we should perhaps move that into Julia itself. It's generated by SnoopCompile, but every package that uses it and has compilable kwbody methods gets its own personal definition.

I'll submit a PR.

**Comment by timholy** [comment]
Note that packages that wish to support versions of Julia older than 1.6 will still need to define that method as is done currently, if you want to precompile the body method. But perhaps easier will be to limit the best precompilation to Julia 1.6, since it will have other advances that improve precompilation.

**Comment by ronisbr** [comment]
I want to use this for TextUserInterfaces.jl. I am already using features that will probably not work in Julia < 1.4 anyway.
